### DISCUSSION TITLE
Optimization

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#19] [Chapter2]
Optimization is worth to consider when your app is grown up. Nobody likes slow app. And you too. Don't you? Memoization can help you to optimize your app.

Memoization makes your computer saves some stuff (Like components, function, object or etc ... 🫡). **So when re-rendering is happened, computer can get saved values. No need to make it twice. But remembering It's not for free.** You should compare which is expensive making it one more time or remembering it.

But one thing! There is a recommanded case for memoization. **When you use `useEffect`, dependency array together, You should better make each element of array as `useCallback` or `useMemo`.** Because objects are saved as a reference as you know. So `useEffect` has posibility for malfunction.

### According to book, the writer said 'Cover all stuffs `useMemo` or `useCallback` as you can. After app has grown up, Calculate which `memoization` should be removed.'

![optimize](https://github.com/honaseung/react-deep-dive/assets/53066347/2705ea15-1544-42fd-8c87-bedc40f822fb)

**Comment by honaseung** [comment]
최적화는 앱이 커졌을 때 고려할 가치가 있습니다. 아무도 느린 앱을 좋아하지 않습니다. 당신도 마찬가지입니다. 그렇지 않나요? 메모이제이션은 앱을 최적화하는 데 도움을 줄 수 있습니다.

메모이제이션을 하면 컴퓨터에 정보(예: 구성 요소, 기능, 개체 등... 🫡)가 저장됩니다. 따라서 리렌더링이 발생하면 컴퓨터는 두 번 만들지 않고 저장된 값을 가져올 수 있습니다. 하지만 저장의 대가는 무료가 아닙니다. 한 번 더 만드는 것과 기억하기 중 어느 것이 비용이 많이 드는지 비교해야 합니다.

그런데 한 가지! 메모이제이션을 추천하는 경우가 있습니다.  useEffect, 의존성 배열을 같이 사용할 때는 배열의 각 요소를 useCallback으로 만들거나 useMemo로 만드는 것이 좋습니다. 왜냐하면 객체는 아시다시피 참조로 저장되기 때문에 useEffect는 오작동 가능성이 있습니다.

책에 따르면 글쓴이는 '모든 것을 useMemo로 덮거나 가능한 한 useCallback을 사용하십시오. 그리고 앱이 커진 후 어떤 메모를 제거해야 하는지 계산하십시오.'라고 말했습니다

**Reply by chaminju927** [reply]
'그런데 한 가지! 메모이제이션을 추천하는 경우가 있습니다. useEffect, 의존성 배열을 같이 사용할 때는 배열의 각 요소를 useCallback으로 만들거나 useMemo로 만드는 것이 좋습니다. 왜냐하면 객체는 아시다시피 참조로 저장되기 때문에 useEffect는 오작동 가능성이 있습니다.' --> 이 부분 정말 공감합니다. 실제로 코드를 작성할때도  useCallback, useMemo로 메모이제이션한 것들을 useEffect 의존성 배열에 넣어야 오류가 잘 안뜨더라구요 !!
